CONTRIBUTING.md

```diff
@@ -209,3 +209,146 @@ HOST=$(kubectl get svc ${RELEASE_NAME}-ingress-nginx-controller --namespace $NAM
+- It is an opportunity to remove legacy compatibility code when a new minor version is released, e.g. when we release `0.12.0`, we can [remove the legacy compatibility code](#removing-compatibility-code) for `0.10.0` in our codebase.
+
+The general guideline to keep backward compatibility is to bump the `API_VERSION` constant in [`sky/server/constants.py`](https://github.com/skypilot-org/skypilot/blob/master/sky/server/constants.py) when introducing an API change and handle backward compatibility based on the `API_VERSION` of the remote peer:
+
+```diff
+# sky/server/constants.py
+- API_VERSION = 11
++ API_VERSION = 12
+
+# Application code
++ from sky.server import versions
++ # Handle the case where the remote peer runs in an API version older than 12
++ if versions.get_remote_api_version() < 12:
++   ...
+```
+
+Some concrete examples are listed below.
+
+#### Adding new APIs
+
+Bump the `API_VERSION` when adding a new API. Then:
+
+- For new SDK methods that calls the new API, add the `@versions.minimal_api_version(API_VERSION)` decorator to the method:
+
+    ```python
+    from sky.server import versions
+
+    @versions.minimal_api_version(12)
+    def new_feature_method():
+        """This method requires server API version 12 or higher."""
+        pass
+    ```
+
+- For existing SDK methods that will be modified to call the new API, the business logic should handle backward compatibility based on the server's API version:
+
+    ```python
+    from sky.server import versions
+
+    def existing_sdk_method():
+        if versions.get_remote_api_version() >= 12:
+            # Call the new API
+        else:
+            # Proceed without the new API. Usually we just keep the same behavior
+            # as before the new API is introduced.
+    ```
+
+#### Adding new fields to API payload
+
+By convention, we define API payloads in `sky/server/api/payloads.py` and there are test cases to enforce the newly added fields must have a default value to keep backward compatibility at API level:
+
+- When receiving a payload from an older version without the new field, the default value is used for the missing new field.
+- When receiving a payload from a newer version with a new field, the value of the new field is ignored.
+
+However, when the value of the new field is taken from an user input (e.g. CLI flag), we should add a warning message to inform the user that the new field is ignored. An API version bump is required in this case. For example:
+
+```python
+from sky.server import versions
+
+@click.option('--newflag', default=None)
+def cli_entry_point(newflag: Optional[str] = None):
+    # The new flag is set but the server does not support the new field yet
+    if newflag is not None and versions.get_remote_api_version() < 12:
+        logger.warning('The new flag is ignored because the server does not support it yet.')
```

Comment:
Hmm, this `get_remote_api_version` may not have been set when the CLI call it like this before any health check has been run?

Reply:
Yes, I will add a note that API server related SDK method must have a `@server_common.check_server_healthy_or_start` decorator added